### ISSUE TITLE
PROBLEMS pane: file path is not correctly prefixed with forward slash `/`

### ISSUE THREAD

**Issue by zfields** [microsoft/vscode-cpptools#1915]
Forwarded from: https://github.com/Microsoft/vscode/issues/48841

**Example output from file pane:**
```
src_file.cpp Users/zfields/dev/source/application/lib/src_file.cpp
* use of undeclared identifier 'id' (213,5)
```

When I click on the problem, I get the following error message.

**Error message:**
```
Unable to open 'src_file.cpp': File not found
(file:///Users/zfields/dev/source/application/Users/zfields/dev/source/application/lib/src_file.cpp).
```

It looks like **VSCode** is trying to use a relative path from `${workspaceFolder}`, even though the PROBLEMS pane is showing an absolute-ish path. Perhaps the path from the PROBLEMS pane would be appropriately identified as an absolute path if it was prefixed with a forward slash `/`.

**Comment by sean-mcmanus** [comment]
The messages "Unable to open … File not found" and "use of undeclared identifier" are not from our extension. Do you have another C++ extension installed? This doesn't sound like our issue, right?

**Comment by zfields** [comment]
First, please don't close issues in less than 24 hours without giving me an opportunity to respond.

Here is the extension I have installed - no other cpp extension.

<img width="320" alt="screen shot 2018-05-03 at 10 04 10 am" src="https://user-images.githubusercontent.com/2801887/39584761-d2ecdf06-4eb8-11e8-801e-4fa81f20dfd4.png">

I had already raised an [issue](https://github.com/Microsoft/vscode/issues/48841) with the VSCode team first, and they directed me to you.

> jrieken:
> It's the extension that feeds those uris into the system. @zfields please forward this to the cpp-extension. Thanks

Perhaps it isn't your error message, but instead you have a poorly constructed URI. Therefore, the file path would in turn be handed off to some module that would generate such an error message.

Is your extension responsible for constructing the URIs?

**Comment by bobbrow** [comment]
If the message doesn't come from our extension, then we didn't construct the URI.  Can you provide more details as to how this message was generated?  Were you running a task?  Were you editing code?  What other extensions have you installed?

**Comment by sean-mcmanus** [comment]
Our extension has a problem matcher, which assumes the path is relative to your workspace as you describe. It's currently "by design". We could treat this as a feature request to enable this to work with absolute paths (no one else has asked for this yet though). You could modify the package.json yourself as a workaround for your scenario, by changing the fileLocation to "absolute" (see https://code.visualstudio.com/docs/editor/tasks-appendix ). I'm not sure yet if there's a way to get it detect the absolute path and switch to the correct problem matcher. The code is in our package.json, search for problemMatchers: (https://github.com/Microsoft/vscode-cpptools/blob/master/Extension/package.json ). This part of the extension is open source.

**Comment by sean-mcmanus** [comment]
The ability to auto-detect if a problem path is absolute or relative is being tracked by https://github.com/Microsoft/vscode/issues/449 . I have upvoted the issue.

However, it sounds like we might be able to workaround the issue via adding an absolute path problem matcher before the relative path one.

**Comment by zfields** [comment]
These are all the extensions I have installed...

<img width="328" alt="screen shot 2018-05-03 at 2 29 37 pm" src="https://user-images.githubusercontent.com/2801887/39598341-d6a78cfc-4edd-11e8-8cea-29dcbfd7ad42.png">

The message is generated when I run a task of `"type": "shell"`, using the `"problemMatcher": ["$gcc"]`.

I'm running a python script that compiles my project using a makefile.

Sorry for the delay, I've been trying to collect better examples, but they are very specific to my work and I haven't found any that are completely free of IP.

**Comment by zfields** [comment]
@sean-mcmanus I believe your intuition is right, that they appear to be related to the problem matcher.

**Comment by hzcheng** [comment]
It seems like when we use "make -C" to make on top level, the problem shows. Any update on this problem?

**Comment by merlijn-sebrechts** [comment]
Note: if you can change the GCC commands themselves, you can fix this issue by using only absolute paths in GCC calls.

```json
                "command": "g++",
                "args": [
                    "-g",
                    "-Wall",
                    // When GCC displays errors, it uses the paths you supply it to show the 
                    // file or the error, even if those paths are relative. However, GCC uses
                    // absolute paths for errors in system libraries.
                    // VSCode however, requires all paths in errors to be either relative OR absolute.
                    // so we append ${workspaceRoot} to the relative paths in this command so that all
                    // paths are absolute.
                    // Note: this requires configuring problemmatcher to use absolute paths too.
                    "${workspaceRoot}/*.cpp",
                    "-static-libgcc",
                    "-static-libstdc++",
                    "-lallegro",
                    "-lallegro_audio",
                    "-lallegro_acodec",
                    "-lallegro_dialog",
                    "-lallegro_font",
                    "-lallegro_image",
                    "-lallegro_primitives",
                    "-lallegro_ttf",
                    "-o",
                    "project.exe"
                ]
```

and then change the problemmatcher to use absolute paths.

```json
            "problemMatcher": {
                "base": "$gcc",
                "fileLocation": "absolute"
            },
```

**Comment by AlexandreBossard** [comment]
Any update on this?

 I have a c++ project which uses conan, cmake and ninja. When configured and built from a subdirectory, for example `${workspaceFolder}/build`, all the path given by `gcc` are of the form '../path/to/file.cpp'. vs code resolves this relative paths from the ${workspaceFolder}, which does not work. Changing the `fileLocation` does not help. This seem to me a normal workflow, (building from a subdirectory), is there a workaround ?

**Comment by bobbrow** [comment]
Fixed via #3916

**Comment by sean-mcmanus** [comment]
But the fix won't be available to try until 0.28.0-insiders is released.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/0.28.0-insiders

**Comment by elahehrashedi** [comment]
This issue is fixed in 0.28.0.
0.28.0 release: https://github.com/microsoft/vscode-cpptools/releases/tag/0.28.0
